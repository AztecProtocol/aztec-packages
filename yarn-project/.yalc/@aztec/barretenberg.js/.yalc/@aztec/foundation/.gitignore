dest
node_modules
.cache
.env
*.log
*.swp
.tsbuildinfo
.pnp.cjs
.pnp.loader.mjs