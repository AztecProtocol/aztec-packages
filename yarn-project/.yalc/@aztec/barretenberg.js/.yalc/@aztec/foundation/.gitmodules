[submodule "build-system"]
	path = build-system
	url = git@github.com:AztecProtocol/build-system.git
