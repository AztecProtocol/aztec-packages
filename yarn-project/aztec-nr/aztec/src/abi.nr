use crate::constants_gen::{
    RETURN_VALUES_LENGTH,
    MAX_READ_REQUESTS_PER_CALL,
    MAX_PENDING_READ_REQUESTS_PER_CALL,
    MAX_NEW_COMMITMENTS_PER_CALL,
    MAX_NEW_NULLIFIERS_PER_CALL,
    MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL,
    MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL,
    MAX_NEW_L2_TO_L1_MSGS_PER_CALL,
    NUM_FIELDS_PER_SHA256,
    MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL,
    MAX_PUBLIC_DATA_READS_PER_CALL,
    GENERATOR_INDEX__FUNCTION_ARGS,
    HISTORIC_BLOCK_DATA_LENGTH,
    CONTRACT_DEPLOYMENT_DATA_LENGTH,
    CALL_CONTEXT_LENGTH,
    PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH,
    PRIVATE_CIRCUIT_PUBLIC_INPUTS_HASH_INPUT_LENGTH,
    CONTRACT_STORAGE_UPDATE_REQUEST_LENGTH,
    CONTRACT_STORAGE_READ_LENGTH,
    PUBLIC_CIRCUIT_PUBLIC_INPUTS_LENGTH,
    PUBLIC_CIRCUIT_PUBLIC_INPUTS_HASH_INPUT_LENGTH,
    GENERATOR_INDEX__BLOCK_HASH,
    GENERATOR_INDEX__PUBLIC_CIRCUIT_PUBLIC_INPUTS,
    GENERATOR_INDEX__FUNCTION_DATA,
    GENERATOR_INDEX__PUBLIC_DATA_READ,
    GENERATOR_INDEX__PUBLIC_DATA_UPDATE_REQUEST,
    GENERATOR_INDEX__CALL_CONTEXT,
    GENERATOR_INDEX__PRIVATE_CIRCUIT_PUBLIC_INPUTS,
    GENERATOR_INDEX__CONTRACT_DEPLOYMENT_DATA,
};

use crate::oracle::debug_log;
use crate::types::vec::BoundedVec;
use crate::types::point::Point;
use crate::hash::pedersen_hash;

// docs:start:private-global-variables
struct PrivateGlobalVariables {
    chain_id: Field,
    version: Field,
}
// docs:end:private-global-variables

impl PrivateGlobalVariables {
    fn serialize(self) -> [Field; 2] {
        [self.chain_id, self.version]
    }
}

// docs:start:public-global-variables
struct PublicGlobalVariables {
    chain_id: Field,
    version: Field,
    block_number: Field,
    timestamp: Field,
}
// docs:end:public-global-variables

impl PublicGlobalVariables {
    fn serialize(self) -> [Field; 4] {
        [self.chain_id, self.version, self.block_number, self.timestamp]
    }
}

// docs:start:contract-deployment-data
struct ContractDeploymentData {
    deployer_public_key: Point,
    constructor_vk_hash : Field,
    function_tree_root : Field,
    contract_address_salt : Field,
    portal_contract_address : Field,
}
// docs:end:contract-deployment-data

impl ContractDeploymentData {
    fn serialize(self) -> [Field; CONTRACT_DEPLOYMENT_DATA_LENGTH] {
        [
            self.deployer_public_key.x,
            self.deployer_public_key.y,
            self.constructor_vk_hash,
            self.function_tree_root,
            self.contract_address_salt,
            self.portal_contract_address,
        ]
    }

    fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__CONTRACT_DEPLOYMENT_DATA)
    }
}

// PrivateContextInputs are expected to be provided to each private function
// docs:start:private-context-inputs
struct PrivateContextInputs {
    call_context : CallContext,
    block_data: HistoricBlockData,
    contract_deployment_data: ContractDeploymentData,
    private_global_variables: PrivateGlobalVariables,
}
// docs:end:private-context-inputs

// PublicContextInputs are expected to be provided to each public function
// docs:start:public-context-inputs
struct PublicContextInputs {
    call_context: CallContext,
    block_data: HistoricBlockData,

    public_global_variables: PublicGlobalVariables,
}
// docs:end:public-context-inputs

// docs:start:call-context
struct CallContext {
    msg_sender : Field,
    storage_contract_address : Field,
    portal_contract_address : Field,
    function_selector: Field,

    is_delegate_call : bool,
    is_static_call : bool,
    is_contract_deployment: bool,
}
// docs:end:call-context

impl CallContext {
    fn serialize(self) -> [Field; CALL_CONTEXT_LENGTH] {
        [
            self.msg_sender,
            self.storage_contract_address,
            self.portal_contract_address,
            self.function_selector,
            self.is_delegate_call as Field,
            self.is_static_call as Field,
            self.is_contract_deployment as Field,
        ]
    }

    fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__CALL_CONTEXT)
    }
}

// docs:start:historic-block-data
struct HistoricBlockData {
    note_hash_tree_root : Field,
    nullifier_tree_root : Field,
    contract_tree_root : Field,
    l1_to_l2_messages_tree_root : Field,
    blocks_tree_root: Field,
    public_data_tree_root: Field,
    global_variables_hash: Field,
}
// docs:end:historic-block-data

impl HistoricBlockData {
    // NOTE: this order must match the order in `private_circuit_public_inputs.hpp`
    pub fn serialize(self) -> [Field; HISTORIC_BLOCK_DATA_LENGTH] {
        [
            self.note_hash_tree_root,
            self.nullifier_tree_root,
            self.contract_tree_root,
            self.l1_to_l2_messages_tree_root,
            self.blocks_tree_root,
            self.public_data_tree_root,
            self.global_variables_hash,
        ]
    }

    pub fn deserialize(deserialized: [Field; HISTORIC_BLOCK_DATA_LENGTH]) -> Self {
        HistoricBlockData {
            note_hash_tree_root: deserialized[0],
            nullifier_tree_root: deserialized[1],
            contract_tree_root: deserialized[2],
            l1_to_l2_messages_tree_root: deserialized[3],
            blocks_tree_root: deserialized[4],
            public_data_tree_root: deserialized[5],
            global_variables_hash: deserialized[6],
        }
    }

    pub fn empty() -> Self {
        Self { note_hash_tree_root: 0, nullifier_tree_root: 0, contract_tree_root: 0, l1_to_l2_messages_tree_root: 0, blocks_tree_root: 0, public_data_tree_root: 0, global_variables_hash: 0 }
    }

    pub fn block_hash(self) -> Field {
        // TODO: Unify the ordering in `HistoricBlockData::serialize` function and the ordering
        // in the block hash preimage --> This requires changes in the circuits.
        let inputs = [
            self.global_variables_hash,
            self.note_hash_tree_root,
            self.nullifier_tree_root,
            self.contract_tree_root,
            self.l1_to_l2_messages_tree_root,
            self.public_data_tree_root
        ];
        pedersen_hash(inputs, GENERATOR_INDEX__BLOCK_HASH)
    }
}

struct FunctionData {
    function_selector: Field,
    is_internal: bool,
    is_private: bool,
    is_constructor: bool,
}

impl FunctionData {
    fn hash(self) -> Field {
        pedersen_hash([
            self.function_selector,
            self.is_internal as Field,
            self.is_private as Field,
            self.is_constructor as Field,
        ], GENERATOR_INDEX__FUNCTION_DATA)
    }
}

struct PrivateCircuitPublicInputs {
    call_context: CallContext,
    args_hash: Field,
    return_values: [Field; RETURN_VALUES_LENGTH],
    read_requests: [Field; crate::abi::MAX_READ_REQUESTS_PER_CALL],
    pending_read_requests: [Field; crate::abi::MAX_PENDING_READ_REQUESTS_PER_CALL],
    new_commitments: [Field; MAX_NEW_COMMITMENTS_PER_CALL],
    new_nullifiers: [Field; MAX_NEW_NULLIFIERS_PER_CALL],
    nullified_commitments: [Field; MAX_NEW_NULLIFIERS_PER_CALL],
    private_call_stack: [Field; MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL],
    public_call_stack: [Field; MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL],
    new_l2_to_l1_msgs: [Field; MAX_NEW_L2_TO_L1_MSGS_PER_CALL],
    // Explore introducing a new type like uint256 (similar to Point), so it's more explicit that
    // we're talking about a single number backed by two field elements.
    encrypted_logs_hash: [Field; NUM_FIELDS_PER_SHA256],
    unencrypted_logs_hash: [Field; NUM_FIELDS_PER_SHA256],
    encrypted_log_preimages_length: Field,
    unencrypted_log_preimages_length: Field,
    block_data: HistoricBlockData,
    contract_deployment_data: ContractDeploymentData,
    chain_id: Field,
    version: Field,
}

impl PrivateCircuitPublicInputs {
    fn hash(self) -> Field {
        let mut fields: BoundedVec<Field, PRIVATE_CIRCUIT_PUBLIC_INPUTS_HASH_INPUT_LENGTH> = BoundedVec::new(0); 
        fields.push(self.call_context.hash());
        fields.push(self.args_hash);
        fields.push_array(self.return_values);
        fields.push_array(self.read_requests);
        fields.push_array(self.pending_read_requests);
        fields.push_array(self.new_commitments);
        fields.push_array(self.new_nullifiers);
        fields.push_array(self.nullified_commitments);
        fields.push_array(self.private_call_stack);
        fields.push_array(self.public_call_stack);
        fields.push_array(self.new_l2_to_l1_msgs);
        fields.push_array(self.encrypted_logs_hash);
        fields.push_array(self.unencrypted_logs_hash);
        fields.push(self.encrypted_log_preimages_length);
        fields.push(self.unencrypted_log_preimages_length);
        fields.push_array(self.block_data.serialize());
        fields.push(self.contract_deployment_data.hash());
        fields.push(self.chain_id);
        fields.push(self.version);

        pedersen_hash(fields.storage, GENERATOR_INDEX__PRIVATE_CIRCUIT_PUBLIC_INPUTS)
    }

    fn serialize(self) -> [Field; PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH] {
        let mut fields: BoundedVec<Field, PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH> = BoundedVec::new(0); 
        fields.push_array(self.call_context.serialize());
        fields.push(self.args_hash);
        fields.push_array(self.return_values);
        fields.push_array(self.read_requests);
        fields.push_array(self.pending_read_requests);
        fields.push_array(self.new_commitments);
        fields.push_array(self.new_nullifiers);
        fields.push_array(self.private_call_stack);
        fields.push_array(self.public_call_stack);
        fields.push_array(self.new_l2_to_l1_msgs);
        fields.push_array(self.encrypted_logs_hash);
        fields.push_array(self.unencrypted_logs_hash);
        fields.push(self.encrypted_log_preimages_length);
        fields.push(self.unencrypted_log_preimages_length);
        fields.push_array(self.block_data.serialize());
        fields.push_array(self.contract_deployment_data.serialize());
        fields.push(self.chain_id);
        fields.push(self.version);
        fields.storage
    }
}

struct ContractStorageRead {
    storage_slot: Field,
    value: Field,
}

impl ContractStorageRead {
    pub fn serialize(self) -> [Field; CONTRACT_STORAGE_READ_LENGTH] {
        [self.storage_slot, self.value]
    }

    pub fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__PUBLIC_DATA_READ)
    }

    pub fn empty() -> Self {
        Self { storage_slot: 0, value: 0 }
    }
}

struct ContractStorageUpdateRequest {
    storage_slot: Field,
    old_value: Field,
    new_value: Field,
}

impl ContractStorageUpdateRequest {
    pub fn serialize(self) -> [Field; CONTRACT_STORAGE_UPDATE_REQUEST_LENGTH] {
        [self.storage_slot, self.old_value, self.new_value]
    }

    pub fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__PUBLIC_DATA_UPDATE_REQUEST)
    }

    pub fn empty() -> Self {
        Self { storage_slot: 0, old_value: 0, new_value: 0 }
    }
}

struct PublicCircuitPublicInputs {
    call_context: CallContext,
    args_hash: Field,
    return_values: [Field; RETURN_VALUES_LENGTH],
    contract_storage_update_requests: [ContractStorageUpdateRequest; MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL],
    contract_storage_read: [ContractStorageRead; MAX_PUBLIC_DATA_READS_PER_CALL],
    public_call_stack: [Field; MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL],
    new_commitments: [Field; MAX_NEW_COMMITMENTS_PER_CALL],
    new_nullifiers: [Field; crate::abi::MAX_NEW_NULLIFIERS_PER_CALL],
    new_l2_to_l1_msgs: [Field; crate::abi::MAX_NEW_L2_TO_L1_MSGS_PER_CALL],
    unencrypted_logs_hash: [Field; NUM_FIELDS_PER_SHA256],
    unencrypted_log_preimages_length: Field,
    block_data: HistoricBlockData,
    prover_address: Field,
}

impl PublicCircuitPublicInputs {
    
    pub fn hash(self) -> Field {
        let mut inputs: BoundedVec<Field, PUBLIC_CIRCUIT_PUBLIC_INPUTS_HASH_INPUT_LENGTH> = BoundedVec::new(0);
        inputs.push(self.call_context.hash());
        inputs.push(self.args_hash);
        inputs.push_array(self.return_values);
        for i in 0..MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL {
            inputs.push(self.contract_storage_update_requests[i].hash());
        }
        for i in 0..MAX_PUBLIC_DATA_READS_PER_CALL {
            inputs.push(self.contract_storage_read[i].hash());
        }
        inputs.push_array(self.public_call_stack);
        inputs.push_array(self.new_commitments);
        inputs.push_array(self.new_nullifiers);
        inputs.push_array(self.new_l2_to_l1_msgs);

        inputs.push_array(self.unencrypted_logs_hash);
        inputs.push(self.unencrypted_log_preimages_length);
        inputs.push_array(self.block_data.serialize());
        inputs.push(self.prover_address);

        pedersen_hash(inputs.storage, GENERATOR_INDEX__PUBLIC_CIRCUIT_PUBLIC_INPUTS)
    }

    pub fn serialize(self) -> [Field; PUBLIC_CIRCUIT_PUBLIC_INPUTS_LENGTH] {
        let mut fields: BoundedVec<Field, PUBLIC_CIRCUIT_PUBLIC_INPUTS_LENGTH> = BoundedVec::new(0); 
        fields.push_array(self.call_context.serialize());   
        fields.push(self.args_hash);
        fields.push_array(self.return_values);
        for i in 0..MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_CALL {
            fields.push_array(self.contract_storage_update_requests[i].serialize());
        }
        for i in 0..MAX_PUBLIC_DATA_READS_PER_CALL {
            fields.push_array(self.contract_storage_read[i].serialize());
        }
        fields.push_array(self.public_call_stack);
        fields.push_array(self.new_commitments);
        fields.push_array(self.new_nullifiers);
        fields.push_array(self.new_l2_to_l1_msgs);
        fields.push_array(self.unencrypted_logs_hash);
        fields.push(self.unencrypted_log_preimages_length);
        fields.push_array(self.block_data.serialize());
        fields.push(self.prover_address);
        fields.storage
    }
}

struct Hasher {
    fields: [Field],
}

impl Hasher {
    pub fn new()-> Self {
        Self { fields: [] }
    }

    pub fn add(&mut self, field: Field) {
        self.fields = self.fields.push_back(field);
    }

    pub fn add_multiple<N>(&mut self, fields: [Field; N]) {
        for i in 0..N {
            self.fields = self.fields.push_back(fields[i]);
        }
    }

    pub fn hash(self) -> Field {
        hash_args(self.fields)
    }
}

global ARGS_HASH_CHUNK_LENGTH: u32 = 32;
global ARGS_HASH_CHUNK_COUNT: u32 = 16;

pub fn hash_args<N>(args: [Field; N]) -> Field {
    if args.len() == 0 {
        0
    } else {
        let mut chunks_hashes = [0; ARGS_HASH_CHUNK_COUNT];
        for i in 0..ARGS_HASH_CHUNK_COUNT {
            let mut chunk_hash = 0;
            let start_chunk_index = i * ARGS_HASH_CHUNK_LENGTH;
            if start_chunk_index < (args.len() as u32) {
                let mut chunk_args = [0; ARGS_HASH_CHUNK_LENGTH];
                for j in 0..ARGS_HASH_CHUNK_LENGTH {
                    let item_index = i * ARGS_HASH_CHUNK_LENGTH + j;
                    if item_index < (args.len() as u32) {
                        chunk_args[j] = args[item_index];
                    }
                }
                chunk_hash = pedersen_hash(chunk_args, GENERATOR_INDEX__FUNCTION_ARGS);
            }
            chunks_hashes[i] = chunk_hash;
        }
        pedersen_hash(chunks_hashes, GENERATOR_INDEX__FUNCTION_ARGS)
    }
}
