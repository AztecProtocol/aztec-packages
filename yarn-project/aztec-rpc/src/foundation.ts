import nodeCrypto from 'crypto';
import { Keccak } from 'sha3';

export const randomBytes = (len: number) => {
  return nodeCrypto.randomBytes(len) as Buffer;
};

export function keccak256(input: Buffer | string) {
  const inputBuf = typeof input === 'string' ? Buffer.from(input) : input;
  const hash = new Keccak(256);
  return hash.update(inputBuf).digest();
}
