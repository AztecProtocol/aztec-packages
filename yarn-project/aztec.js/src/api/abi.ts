export { ContractArtifact, FunctionArtifact, FunctionSelector } from '@aztec/foundation/abi';
export { loadContractArtifact } from '@aztec/types/abi';
export { NoirCompiledContract } from '@aztec/types/noir';
