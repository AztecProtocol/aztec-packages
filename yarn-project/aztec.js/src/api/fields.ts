export { Point, Fr, Fq, GrumpkinScalar } from '@aztec/foundation/fields';
