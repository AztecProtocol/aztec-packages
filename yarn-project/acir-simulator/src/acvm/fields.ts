import { AztecAddress, EthAddress, Fr } from '@aztec/foundation';

export type ACVMField = `0x${string}`;
export type ACVMWitness = Map<number, ACVMField>;

export const ZERO_ACVM_FIELD: ACVMField = `0x${Buffer.alloc(32).toString('hex')}`;
export const ONE_ACVM_FIELD: ACVMField = `0x${'00'.repeat(31)}01`;

function adaptBufferSize(originalBuf: Buffer) {
  const buffer = Buffer.alloc(32);
  if (originalBuf.length > buffer.length) {
    throw new Error('Buffer does not fit in 32 bytes');
  }
  originalBuf.copy(buffer, buffer.length - originalBuf.length);
  return buffer;
}

export function toACVMField(value: AztecAddress | EthAddress | Fr | Buffer | boolean | number): `0x${string}` {
  if (typeof value === 'boolean') {
    return value ? ONE_ACVM_FIELD : ZERO_ACVM_FIELD;
  }

  let buffer;

  if (Buffer.isBuffer(value)) {
    buffer = value;
  } else if (typeof value === 'number') {
    buffer = Buffer.alloc(32);
    buffer.writeUInt32BE(value, 28);
  } else {
    buffer = value.toBuffer();
  }

  return `0x${adaptBufferSize(buffer).toString('hex')}`;
}

export function fromACVMField(field: `0x${string}`): Fr {
  const buffer = Buffer.from(field.slice(2), 'hex');
  return Fr.fromBuffer(buffer);
}
