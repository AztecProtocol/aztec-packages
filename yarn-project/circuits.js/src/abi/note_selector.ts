import { toBigIntBE } from '@aztec/foundation/bigint-buffer';
import { randomInt } from '@aztec/foundation/crypto';
import { Fr } from '@aztec/foundation/fields';
import { hexSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, TypeRegistry } from '@aztec/foundation/serialize';

import { Selector } from './selector.js';

/* eslint-disable @typescript-eslint/no-unsafe-declaration-merging */

/** Note selector branding */
export interface NoteSelector {
  /** Brand. */
  _branding: 'NoteSelector';
}

/**
 * A note selector is a 7 bit long value that identifies a note type within a contract.
 * TODO(#10952): Encoding of note type id can be reduced to 7 bits.
 */
export class NoteSelector extends Selector {
  /**
   * Deserializes from a buffer or reader, corresponding to a write in cpp.
   * @param buffer - Buffer  or BufferReader to read from.
   * @returns The Selector.
   */
  static fromBuffer(buffer: Buffer | BufferReader) {
    const reader = BufferReader.asReader(buffer);
    const value = Number(toBigIntBE(reader.readBytes(Selector.SIZE)));
    if (value >= 1 << 7) {
      throw new Error(`Invalid note selector: ${value}`);
    }
    return new NoteSelector(value);
  }

  static fromString(buf: string) {
    const withoutPrefix = buf.replace(/^0x/i, '').slice(-8);
    const buffer = Buffer.from(withoutPrefix, 'hex');
    return NoteSelector.fromBuffer(buffer);
  }

  /**
   * Converts a field to selector.
   * @param fr - The field to convert.
   * @returns The selector.
   */
  static fromField(fr: Fr) {
    return new NoteSelector(Number(fr.toBigInt()));
  }

  /**
   * Creates an empty selector.
   * @returns An empty selector.
   */
  static empty() {
    return new NoteSelector(0);
  }

  /**
   * Creates a random selector.
   * @returns A random selector.
   */
  static random() {
    const value = randomInt(1 << 7);
    return NoteSelector.fromField(new Fr(value));
  }

  toJSON() {
    return this.toString();
  }

  static get schema() {
    return hexSchemaFor(NoteSelector);
  }
}

// For deserializing JSON.
TypeRegistry.register('NoteSelector', NoteSelector);
