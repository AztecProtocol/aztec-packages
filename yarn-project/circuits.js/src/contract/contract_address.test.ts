import { ABIParameterVisibility, FunctionAbi, FunctionType } from '@aztec/foundation/abi';
import { Fr, Point } from '@aztec/foundation/fields';

import { EthAddress } from '../index.js';
import {
  computeContractAddressFromInstance,
  computeContractAddressFromPartial,
  computeInitializationHash,
  computePartialAddress,
  computePublicKeysHash,
  computeSaltedInitializationHash,
} from './contract_address.js';

describe('ContractAddress', () => {
  it('computePartialAddress', () => {
    const mockInstance = {
      contractClassId: new Fr(1),
      saltedInitializationHash: new Fr(2),
    };
    const result = computePartialAddress(mockInstance);
    expect(result).toMatchSnapshot();
  });

  it('computeSaltedInitializationHash', () => {
    const mockInstance = {
      initializationHash: new Fr(1),
      salt: new Fr(2),
      portalContractAddress: EthAddress.fromField(new Fr(3)),
    };
    const result = computeSaltedInitializationHash(mockInstance);
    expect(result).toMatchSnapshot();
  });

  it('computeInitializationHash', () => {
    const mockInitFn: FunctionAbi = {
      functionType: FunctionType.SECRET,
      isInternal: false,
      name: 'fun',
      parameters: [{ name: 'param1', type: { kind: 'boolean' }, visibility: ABIParameterVisibility.SECRET }],
      returnTypes: [],
    };
    const mockArgs: any[] = [true];
    const result = computeInitializationHash(mockInitFn, mockArgs);
    expect(result).toMatchSnapshot();
  });

  it('computeContractAddressFromInstance', () => {
    const publicKey = new Point(new Fr(1n), new Fr(2n));
    const salt = new Fr(3n);
    const contractClassId = new Fr(4n);
    const initializationHash = new Fr(5n);
    const portalContractAddress = EthAddress.fromField(new Fr(6n));

    const address = computeContractAddressFromInstance({
      publicKeysHash: computePublicKeysHash(publicKey),
      salt,
      contractClassId,
      initializationHash,
      portalContractAddress,
      version: 1,
    }).toString();

    expect(address).toMatchSnapshot();

    // Value used in "compute_address" test in aztec_address.nr
    // console.log("address", address);
  });

  it('Public key hash matches Noir', () => {
    const publicKey = new Point(new Fr(1n), new Fr(2n));
    const hash = computePublicKeysHash(publicKey).toString();
    expect(hash).toMatchSnapshot();

    // Value used in "compute_public_keys_hash" test in public_keys_hash.nr
    // console.log("hash", hash);
  });

  it('Address from partial matches Noir', () => {
    const publicKey = new Point(new Fr(1n), new Fr(2n));
    const partialAddress = new Fr(3n);
    const address = computeContractAddressFromPartial({ publicKey, partialAddress }).toString();
    expect(address).toMatchSnapshot();

    // Value used in "compute_address_from_partial_and_pubkey" test in aztec_address.nr
    // console.log("address", address);
  });
});
