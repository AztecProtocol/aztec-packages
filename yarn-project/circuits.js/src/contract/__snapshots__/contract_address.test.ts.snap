// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractAddress Address from partial matches Noir 1`] = `"0x0447f893197175723deb223696e2e96dbba1e707ee8507766373558877e74197"`;

exports[`ContractAddress Public key hash matches Noir 1`] = `"0x1923a6246e305720b6aaf751fde0342613e93c82e455c3831e28375c16dd40d8"`;

exports[`ContractAddress computeContractAddressFromInstance 1`] = `"0x2fd71a4f0742364f194dd16d0ae32d2f47845ddc7f5d328f37d4148b565c4123"`;

exports[`ContractAddress computeInitializationHash 1`] = `
Fr {
  "asBigInt": 11287307308183188516369033835241775664908452274022428157981236923904607656063n,
  "asBuffer": {
    "data": [
      24,
      244,
      99,
      184,
      236,
      16,
      42,
      8,
      156,
      101,
      39,
      106,
      125,
      199,
      236,
      87,
      43,
      9,
      28,
      163,
      148,
      181,
      224,
      108,
      3,
      191,
      211,
      120,
      203,
      68,
      24,
      127,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ContractAddress computePartialAddress 1`] = `
Fr {
  "asBigInt": 11370807533904788559065519582196668367978392224620203970943957018221100220632n,
  "asBuffer": {
    "data": [
      25,
      35,
      166,
      36,
      110,
      48,
      87,
      32,
      182,
      170,
      247,
      81,
      253,
      224,
      52,
      38,
      19,
      233,
      60,
      130,
      228,
      85,
      195,
      131,
      30,
      40,
      55,
      92,
      22,
      221,
      64,
      216,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ContractAddress computeSaltedInitializationHash 1`] = `
Fr {
  "asBigInt": 11725758444245911667025964940014811412794639055001970719147766917493953867026n,
  "asBuffer": {
    "data": [
      25,
      236,
      139,
      73,
      109,
      192,
      136,
      17,
      189,
      50,
      237,
      56,
      134,
      9,
      56,
      184,
      97,
      39,
      147,
      221,
      79,
      57,
      195,
      179,
      116,
      155,
      140,
      213,
      210,
      243,
      45,
      18,
    ],
    "type": "Buffer",
  },
}
`;
